<a href="https://colab.research.google.com/github/A01737866/Modelos-de-Regresion-Lineal/blob/main/Modelos_de_Regresi%C3%B3n_Lineal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Importamos librerías
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
#Cargar archivo csv desde equipo
from google.colab import files
files.upload()

In [ ]:
#Cargar desde un archivo .csv sin indice
data = pd.read_csv('cuentas_credicel.csv', encoding='latin')
data

In [ ]:
#Identificar valores nulos por columna
valores_nulos = data.isnull().sum()
valores_nulos

In [ ]:
Filtro = data.iloc[:, [6,7,8,11,12,20,26,27]]
Filtro

In [ ]:
#Identificar valores nulos por columna
valores_nulos = Filtro.isnull().sum()
valores_nulos

In [ ]:
Filtro.info()

In [ ]:
#Conversión de tipo de dato de columna de tipo Object a int
Filtro['plazo']= Filtro['plazo'].astype(int)

In [ ]:
#Eliminar un signo de una columna
Filtro['plazo']=Filtro['plazo'].str.replace('S', '')

In [ ]:
#Compruebo que la columna plazo ya se haya convertido a tipo numérico
Filtro.info()

In [ ]:
#Conversión de tipo de dato de columna de tipo Object a int
Filtro['riesgo']= Filtro['riesgo'].astype(float)

In [ ]:
#Sustituyo valores nulos por el numero "0"
#Tercer método de sustitución de valores nulos
#Sustituir valores nulos por un valor numérico en  concreto
Filtro["riesgo"]= Filtro["riesgo"].fillna(0)

In [ ]:
#Analizar categorias de una columna
riesgo_categorias = Filtro.groupby(['riesgo'])['riesgo'].count()
riesgo_categorias

In [ ]:
#Sustituyo los strings Atraso, Cancelado y Fraude por la constante "0"
Filtro['riesgo']= Filtro['riesgo'].str.replace('Atraso', '0')
Filtro['riesgo']= Filtro['riesgo'].str.replace('Cancelado', '0')
Filtro['riesgo']= Filtro['riesgo'].str.replace('Fraude', '0')

In [ ]:
#Conversión de tipo de dato de columna de tipo Object a int
Filtro['riesgo']= Filtro['riesgo'].astype(float)

In [ ]:
#Compruebo que la columna riesgo ya se halla convertido a tipo numérico
Filtro.info()

In [ ]:
#Identificar valores nulos por columna
valores_nulos = Filtro.isnull().sum()
valores_nulos

In [ ]:
#Analizar categorias de una columna
porc_enganche_categorias = Filtro.groupby(['porc_enganche'])['porc_enganche'].count()
porc_enganche_categorias

In [ ]:
#Sustituir valores nulos con promedio o media
Filtro["porc_enganche"] = Filtro["porc_enganche"].fillna(round(data["porc_enganche"].mean(),1))

In [ ]:
#Identificar valores nulos por columna
valores_nulos = Filtro.isnull().sum()
valores_nulos

**En este caso todas las variables con las que se trabaja son cuantitativas, por lo mismo no se separa el dataframe, es decir, y = Cuantitativas = Filtro**


In [ ]:
#Método aplicando Cuartiles. Encuentro cuartiles 0.25 y 0.75
y= Filtro

percentile25=y.quantile(0.25) #Q1
percentile75=y.quantile(0.75) #Q3
iqr= percentile75 - percentile25

Limite_Superior_iqr= percentile75 + 1.5*iqr
Limite_Inferior_iqr= percentile25 - 1.5*iqr
print("Limite superior permitido", Limite_Superior_iqr)
print("Limite inferior permitido", Limite_Inferior_iqr)

In [ ]:
#Obtenemos datos limpios del Dataframe
Datos_sin_Outliers_iqr= Filtro[(y<=Limite_Superior_iqr)&(y>=Limite_Inferior_iqr)]
Datos_sin_Outliers_iqr

In [ ]:
#Corroboramos valores nulos
valores_nulos=Datos_sin_Outliers_iqr.isnull().sum()
valores_nulos

In [ ]:
#Reemplazamos valores atípicos (nulos) del dataframe con "mean"
#Realizamos una copia del dataframe
data_clean_iqr=Datos_sin_Outliers_iqr.copy()
data_clean_iqr=data_clean_iqr.fillna(round(Datos_sin_Outliers_iqr.mean(),1))
data_clean_iqr

In [ ]:
#Corroboramos valores nulos
valores_nulos=data_clean_iqr.isnull().sum()
valores_nulos

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Enganche) e independiente (Precio)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "precio", y = "enganche", color = "blue", data = data_clean_iqr)

In [161]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["precio"]]
Var_Dep = data_clean_iqr[["enganche"]]

In [162]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [ ]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

**Modelo matemático: y = 0.16992509x + 198.13348551**

**Modelo matemático: Enganche = 0.16992509*Precio + 198.13348551**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
coef_deter

In [167]:
correlation_matrix = data_clean_iqr.corr()
print("Matriz de correlación:")
print(correlation_matrix)

Matriz de correlación:
                     plazo    precio  enganche  monto_financiado  costo_total  \
plazo             1.000000  0.216410  0.080335          0.238428     0.499865   
precio            0.216410  1.000000  0.663595          0.908594     0.797407   
enganche          0.080335  0.663595  1.000000          0.509798     0.458437   
monto_financiado  0.238428  0.908594  0.509798          1.000000     0.853044   
costo_total       0.499865  0.797407  0.458437          0.853044     1.000000   
pagos_realizados  0.169521  0.007190 -0.092871          0.042108     0.093833   
riesgo           -0.001184 -0.007871  0.076431         -0.029810    -0.008648   
porc_enganche    -0.140027 -0.027719  0.215626         -0.112910    -0.113645   

                  pagos_realizados    riesgo  porc_enganche  
plazo                     0.169521 -0.001184      -0.140027  
precio                    0.007190 -0.007871      -0.027719  
enganche                 -0.092871  0.076431       0.215626  

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Precio) e independiente (Monto Financiado)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "monto_financiado", y = "precio", color = "blue", data = data_clean_iqr)

In [176]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["monto_financiado"]]
Var_Dep = data_clean_iqr[["precio"]]

In [177]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [ ]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

**Modelo matemático: y = 1.18232773*x + 386.36664265**

**Modelo matemático: Precio = 1.18232773*Monto Financiado + 386.36664265**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

In [ ]:
correlation_coefficient = data_clean_iqr["precio"].corr(data_clean_iqr["monto_financiado"])

print("Coeficiente de correlación:", correlation_coefficient)

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Costo Total) e independiente (Monto Financiado)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "monto_financiado", y = "costo_total", color = "blue", data = data_clean_iqr)

In [182]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["monto_financiado"]]
Var_Dep = data_clean_iqr[["costo_total"]]

In [183]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [ ]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

**Modelo matemático: y = 1.73195861*x + 249.57571688**

**Modelo matemático: Costo Total = 1.73195861*Monto Financiado + 249.57571688**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

In [ ]:
correlation_coefficient = data_clean_iqr["costo_total"].corr(data_clean_iqr["monto_financiado"])

print("Coeficiente de correlación:", correlation_coefficient)

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Pagos Realizados) e independiente (Porcentaje de Enganche)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "porc_enganche", y = "pagos_realizados", color = "blue", data = data_clean_iqr)

In [190]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["porc_enganche"]]
Var_Dep = data_clean_iqr[["pagos_realizados"]]

In [191]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [ ]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

**Modelo matemático: y = -0.74834627*x + 9.22034857**

**Modelo matemático: Pagos Realizados = -0.74834627*Porcentaje de Enganche + 9.22034857**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

In [ ]:
correlation_coefficient = data_clean_iqr["pagos_realizados"].corr(data_clean_iqr["porc_enganche"])

print("Coeficiente de correlación:", correlation_coefficient)

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Riesgo) e independiente (Porcentaje de Enganche)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "porc_enganche", y = "riesgo", color = "blue", data = data_clean_iqr)

In [198]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["porc_enganche"]]
Var_Dep = data_clean_iqr[["riesgo"]]

In [199]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [201]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

{'fit_intercept': True,
 'copy_X': True,
 'n_jobs': None,
 'positive': False,
 'feature_names_in_': array(['porc_enganche'], dtype=object),
 'n_features_in_': 1,
 'coef_': array([[0.23500076]]),
 'rank_': 1,
 'singular_': array([315.995682]),
 'intercept_': array([0.71231188])}

**Modelo matemático: y = 0.23500076*x + 0.71231188**

**Modelo matemático: Riesgo = 0.23500076*Porcentaje de Enganche + 0.71231188**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

In [ ]:
correlation_coefficient = data_clean_iqr["riesgo"].corr(data_clean_iqr["porc_enganche"])

print("Coeficiente de correlación:", correlation_coefficient)

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Plazo) e independiente (Costo Total)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "costo_total", y = "plazo", color = "blue", data = data_clean_iqr)

In [211]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["costo_total"]]
Var_Dep = data_clean_iqr[["plazo"]]

In [212]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [ ]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

**Modelo matemático: y = 0.00257056*x + 11.90387408**

**Modelo matemático: Plazo = 0.00257056*Costo Total + 11.90387408**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

In [ ]:
correlation_coefficient = data_clean_iqr["plazo"].corr(data_clean_iqr["costo_total"])

print("Coeficiente de correlación:", correlation_coefficient)

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Monto Financiado) e independiente (Precio)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "precio", y = "monto_financiado", color = "blue", data = data_clean_iqr)

In [219]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["precio"]]
Var_Dep = data_clean_iqr[["monto_financiado"]]

In [220]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [ ]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

**Modelo matemático: y = 0.69823501*x + 253.90843778**

**Modelo matemático: Monto Financiado = 0.69823501*Precio + 253.90843778**

In [ ]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

In [ ]:
correlation_coefficient = data_clean_iqr["monto_financiado"].corr(data_clean_iqr["precio"])

print("Coeficiente de correlación:", correlation_coefficient)

In [ ]:
#Imprimimos el scatter plot entre la variable dependiente (Porcentaje de Enganche) e independiente (Enganche)
#Para observar el comportamiento en su dispersión
from turtle import color
sns.scatterplot(x = "enganche", y = "porc_enganche", color = "blue", data = data_clean_iqr)

In [226]:
#Declaramos las variables dependientes e independientes para la regresión lineal
Vars_Indep = data_clean_iqr[["enganche"]]
Var_Dep = data_clean_iqr[["porc_enganche"]]

In [227]:
#Se define model como la función de regresión lineal
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [ ]:
#Ajustamos el modelo con las variables antes declaradas
model.fit(X = Vars_Indep, y = Var_Dep)

In [229]:
#Verificamos los coeficientes obtenidos para el modelo ajustado
model.__dict__

{'fit_intercept': True,
 'copy_X': True,
 'n_jobs': None,
 'positive': False,
 'feature_names_in_': array(['enganche'], dtype=object),
 'n_features_in_': 1,
 'coef_': array([[0.00141764]]),
 'rank_': 1,
 'singular_': array([48063.64212986]),
 'intercept_': array([-1.59051725])}

**Modelo matemático: y = 0.00141764*x - 1.59051725**

**Modelo matemático: Porcentaje de Enganche = 0.00141764*Enganche - 1.59051725**

In [230]:
coef_deter = model.score(X = Vars_Indep, y = Var_Dep)
print("Coeficiente de determinación:", coef_deter)

Coeficiente de determinación: 0.04649454581175594


In [231]:
correlation_coefficient = data_clean_iqr["porc_enganche"].corr(data_clean_iqr["enganche"])

print("Coeficiente de correlación:", correlation_coefficient)

Coeficiente de correlación: 0.215625939561445


In [217]:
correlation_matrix = data_clean_iqr.corr()
print("Matriz de correlación:")
print(correlation_matrix)

Matriz de correlación:
                     plazo    precio  enganche  monto_financiado  costo_total  \
plazo             1.000000  0.216410  0.080335          0.238428     0.499865   
precio            0.216410  1.000000  0.663595          0.908594     0.797407   
enganche          0.080335  0.663595  1.000000          0.509798     0.458437   
monto_financiado  0.238428  0.908594  0.509798          1.000000     0.853044   
costo_total       0.499865  0.797407  0.458437          0.853044     1.000000   
pagos_realizados  0.169521  0.007190 -0.092871          0.042108     0.093833   
riesgo           -0.001184 -0.007871  0.076431         -0.029810    -0.008648   
porc_enganche    -0.140027 -0.027719  0.215626         -0.112910    -0.113645   

                  pagos_realizados    riesgo  porc_enganche  
plazo                     0.169521 -0.001184      -0.140027  
precio                    0.007190 -0.007871      -0.027719  
enganche                 -0.092871  0.076431       0.215626  

In [232]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22735 entries, 0 to 22734
Data columns (total 37 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   folio               22735 non-null  int64  
 1   tag                 22735 non-null  object 
 2   folio_solicitud     22735 non-null  int64  
 3   fecha               22735 non-null  object 
 4   marca               22735 non-null  object 
 5   modelo              22735 non-null  object 
 6   plazo               22735 non-null  object 
 7   precio              22735 non-null  float64
 8   enganche            22735 non-null  float64
 9   descuento           22735 non-null  float64
 10  semana              22735 non-null  int64  
 11  monto_financiado    22735 non-null  float64
 12  costo_total         22735 non-null  int64  
 13  monto_accesorios    22735 non-null  float64
 14  agente_venta        22735 non-null  object 
 15  dis_venta           22735 non-null  object 
 16  stat

In [ ]:
#Filtro para obtener variables cuantitativas
Cuantitativas = data.iloc[:, [0,2,7,8,9,10,11,12,13,16,17,19,20,21,25,27,28,29,35,36]]
Cuantitativas

In [ ]:
#Método aplicando desviación estándar. Encuentro los valores extremos
y = Cuantitativas
Limite_Superior = y.mean() + 3*y.std()
Limite_Inferior = y.mean() - 3*y.std()
print("Límite Superior Permitido", Limite_Superior)
print("Límite Inferior Permitido", Limite_Inferior)

In [235]:
#Ajustar máximo de filas
pd.options.display.max_rows = None

In [ ]:
#Obtenemos datos y los outliers se convierten en nulos en el DataFrame
Datos_sin_Outliers = Cuantitativas[(y <= Limite_Superior) & (y >= Limite_Inferior)]
Datos_sin_Outliers

In [241]:
#Corroboramos valores nulos
valores_nulos = Datos_sin_Outliers.isnull().sum()
valores_nulos

folio                   0
folio_solicitud         0
precio                110
enganche              292
descuento              17
semana                362
monto_financiado      182
costo_total           290
monto_accesorios      617
status                154
fraude               1701
inversion               0
pagos_realizados      184
reautorizacion        774
puntos               4408
porc_enganche        2946
porc_tasa             899
score_buro            582
Unnamed: 35         20404
Unnamed: 36         22735
dtype: int64

In [ ]:
#Reemplazamos valores atípicos (nulos) del dataframe con "mean"
#Realizamos una copia del dataframe
data_clean = Datos_sin_Outliers.copy()
data_clean = data_clean.fillna(round(Datos_sin_Outliers.mean(),1))
data_clean

In [ ]:
#Sustituir valores nulos por un valor numérico en concreto
data_clean["Unnamed: 36"] = data_clean["Unnamed: 36"].fillna(0)
data_clean

In [ ]:
#Corroboramos valores nulos
valores_nulos = data_clean.isnull().sum()
valores_nulos

In [246]:
correlation_matrix = data_clean.corr()
print("Matriz de correlación:")
print(correlation_matrix)

Matriz de correlación:
                     folio  folio_solicitud    precio  enganche  descuento  \
folio             1.000000         0.971332  0.089539  0.125603   0.134939   
folio_solicitud   0.971332         1.000000  0.090329  0.122568   0.131475   
precio            0.089539         0.090329  1.000000  0.690542   0.018582   
enganche          0.125603         0.122568  0.690542  1.000000  -0.025831   
descuento         0.134939         0.131475  0.018582 -0.025831   1.000000   
semana            0.097617         0.099369  0.611028  0.379107  -0.019941   
monto_financiado  0.070554         0.072683  0.935647  0.515344   0.033970   
costo_total       0.048342         0.049955  0.838116  0.448240   0.050457   
monto_accesorios  0.105582         0.094277 -0.024338 -0.015719   0.025154   
status           -0.401955        -0.391637 -0.097610 -0.074212  -0.068609   
fraude                 NaN              NaN       NaN       NaN        NaN   
inversion        -0.139744        -0.1309

In [ ]:
#Método aplicando Cuartiles. Encuentro cuartiles 0.25 y 0.75
y=Cuantitativas

percentile25=y.quantile(0.25) #Q1
percentile75=y.quantile(0.75) #Q3
iqr= percentile75 - percentile25

Limite_Superior_iqr= percentile75 + 1.5*iqr
Limite_Inferior_iqr= percentile25 - 1.5*iqr
print("Limite superior permitido", Limite_Superior_iqr)
print("Limite inferior permitido", Limite_Inferior_iqr)

In [ ]:
#Obtenemos datos limpios del Dataframe
Datos_sin_Outliers_iqr= Cuantitativas[(y<=Limite_Superior_iqr)&(y>=Limite_Inferior_iqr)]
Datos_sin_Outliers_iqr

In [ ]:
#Corroboramos valores nulos
valores_nulos=Datos_sin_Outliers_iqr.isnull().sum()
valores_nulos

In [ ]:
#Reemplazamos valores atípicos (nulos) del dataframe con "mean"
#Realizamos una copia del dataframe
data_clean_iqr=Datos_sin_Outliers_iqr.copy()
data_clean_iqr=data_clean_iqr.fillna(round(Datos_sin_Outliers_iqr.mean(),1))
data_clean_iqr

In [ ]:
#Corroboramos valores nulos
valores_nulos=data_clean_iqr.isnull().sum()
valores_nulos

In [254]:
correlation_matrix = data_clean_iqr.corr()
print("Matriz de correlación:")
print(correlation_matrix)

Matriz de correlación:
                     folio  folio_solicitud    precio  enganche  descuento  \
folio             1.000000         0.971332  0.099368  0.145340   0.130377   
folio_solicitud   0.971332         1.000000  0.099848  0.140092   0.127074   
precio            0.099368         0.099848  1.000000  0.663595   0.025226   
enganche          0.145340         0.140092  0.663595  1.000000  -0.016966   
descuento         0.130377         0.127074  0.025226 -0.016966   1.000000   
semana            0.087642         0.089457  0.573740  0.344543  -0.013835   
monto_financiado  0.090175         0.091934  0.908594  0.509798   0.041332   
costo_total       0.083302         0.083734  0.797407  0.458437   0.063645   
monto_accesorios       NaN              NaN       NaN       NaN        NaN   
status                 NaN              NaN       NaN       NaN        NaN   
fraude                 NaN              NaN       NaN       NaN        NaN   
inversion              NaN              N